internal/app/project/gradle.go

```diff
@@ -1,13 +1,205 @@
+					runTask = fmt.Sprintf(":%s:run", dirName)
+					buildTask = fmt.Sprintf(":%s:classes", dirName)
+					guessed = true
+					break
+				}
+			}
+		}
+	}
+
+	return
+}
+func searchKtorPlugin(dir string) (version semver, found bool) {
+	buildPath := filepath.Join(dir, "build.gradle.kts")
+	buildRoot, err, _ := gradle.ParseBuildFile(buildPath)
+
+	if err != nil {
+		return
+	}
+
+	found = true
+	var catalogKeys []string
+	for _, p := range buildRoot.Plugins.List {
+		if p.IsCatalog && p.Applied {
+			if segments := strings.Split(p.Id, "."); len(segments) > 0 {
+				off := len(segments) - 1
+
+				for ; off >= 0; off-- {
+					if segments[off] == "plugins" {
+						off++
+						break
+					}
+				}
+
+				catalogKeys = append(catalogKeys, strings.Join(segments[off:], "-"))
+			}
+		}
+
+		if p.Prefix == "id" && p.Id == "io.ktor.plugin" && p.Applied {
+			if p.Version != "" {
+				version = parseVersion(p.Version)
+			}
+			return
+		}
+	}
+
+	if len(catalogKeys) > 0 {
+		hasCatalog := false
+		catalogName := "libs.versions.toml"
+		catalogPath := filepath.Join(dir, "gradle", catalogName)
+		if utils.Exists(catalogPath) {
+			hasCatalog = true
+		} else {
+			catalogPath = filepath.Join(dir, "..", "gradle", catalogName)
+
+			if utils.Exists(catalogPath) {
+				hasCatalog = true
+			}
+		}
+
+		if hasCatalog {
+			tomlDoc, err, _ := toml.ParseCatalogToml(catalogPath)
+
+			if err != nil {
+				found = false
+				return
+			}
+
+			if t, ok := toml.FindTable(tomlDoc, "plugins"); ok {
+				for _, te := range t.Entries {
+					if te.Kind == toml.ValueMap {
+						if id, ok := te.Get("id"); ok && id == "io.ktor.plugin" {
+
+							pluginApplied := false
+							for _, catalogKey := range catalogKeys {
+								if te.Key == catalogKey {
+									pluginApplied = true
+									break
+								}
+							}
+
+							if pluginApplied {
+								found = true
+
+								if versionRef, ok := te.Get("version.ref"); ok {
+									if vTable, ok := toml.FindTable(tomlDoc, "versions"); ok {
+										for _, vt := range vTable.Entries {
+											if vt.Key == versionRef && vt.Kind == toml.StringValue {
+												version = parseVersion(vt.String)
+												break
+											}
+										}
+									}
+								}
+
+								return
+							}
+						}
+					}
+				}
+			}
+		}
+	}
+
+	found = false
 	return
 }
+
+func parseVersion(str string) (version semver) {
+	comps := strings.Split(str, ".")
+	if len(comps) == 3 {
+		majorComp, minorComp, patchComp := comps[0], comps[1], comps[2]
+
+		major, err := strconv.Atoi(majorComp)
+		if err != nil {
+			return
+		}
+
+		minor, err := strconv.Atoi(minorComp)
+		if err != nil {
+			return
+		}
+
+		patch := 0
+		for _, c := range patchComp {
+			if unicode.IsNumber(c) { // May be beta or eap
+				patch += patch*10 + (int(c) - '0')
+			} else {
+				break
+			}
+		}
+
+		version = semver{major: major, minor: minor, patch: patch, valid: true}
```

Comment:
Just curious. When `valid` will be `false`? By default?

Reply:
Yes, https://go.dev/tour/basics/12